### PR TITLE
feat: Added lifecycle configuration, bucket owner controls and bucket notification features

### PR DESCRIPTION
**Pull request by patrickmukumbu**
# Description

This PR was created to add lifecycle configuration and bucket notification hence fixing checkov alerts `CKV2_AWS_62` and `CKV2_AWS_61`. Bucket ownership control was also added to fix acl error which resulted from s3 updated by aws in April 2023, [here](https://aws.amazon.com/blogs/aws/heads-up-amazon-s3-security-changes-are-coming-in-april-of-2023/)

## Features/Fixes/Patches/Changes list:

Please check the [CHANGELOG.md](https://github.com/boldlink/terraform-aws-s3/blob/feature/add-lc-config-notifications/CHANGELOG.md#230---2023-09-07)

### PR Owners Checklist:
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you checked there aren't open [Pull Requests](../../../pulls) of upstream or parallel code that will cause conflicts?
* [x] Have you updated the `CHANGELOG.md`?
* [x] Have you updated the `README.md`(s)?
* [x] OWNER: Does your submission pass?
    * [x] Pre-commit (attach logs/print screen to this PR)
    * [x] Checkov/terrascan (attach logs/print screen to this PR)
    * [x] Terraform resources examples build/destroy successfully with Terraform 0.14.11 minimum?
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] First PR? Have you deleted `## How to use this template...` from the root `.README.md`?

### PR Reviewers Checklists?
* [ ] Are all your comments/changes resolved?
* [ ] Are you happy with the OWNER checklists and additional information provided?
* [ ] Does your review tests pass?
    * [ ] Terraform resources examples build/destroy successfully with Terraform 0.14.11 minimum?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -72,6 +76,7 @@ The following features have now been added as stand-alone resources in this rele
 
 [Unreleased]: https://github.com/boldlink/terraform-aws-s3/compare/2.2.0...HEAD
 
+[2.3.0]: https://github.com/boldlink/terraform-aws-s3/releases/tag/2.3.0
```
Comment (hugoalmeida264) [inline CHANGELOG.md:80]: Update the unreleased link to 2.3.0

File: examples/replication_configuration/main.tf [hunk 2]
```diff
@@ -24,18 +24,18 @@ module "replication_role" {
 module "source_kms_key" {
   #checkov:skip=CKV_TF_1: "Ensure Terraform module sources use a commit hash"
   source           = "boldlink/kms/aws"
-  description      = "kms key for ${local.source_bucket}"
+  description      = "kms key for ${var.source_bucket}"
```
Comment (hugoalmeida264) [inline examples/replication_configuration/main.tf]: This should be locked down to a version, if we introduce breaking changes, it breaks the example.

Also, the `CKV_TF_1` should be global, it should be on `.checkov.yaml` file

File: README.md [hunk 3]
```diff
@@ -16,10 +16,15 @@
 ## Description
 This terraform module creates an S3 Bucket with the options of adding other s3 bucket configurations.
 
-Examples available [here](https://github.com/boldlink/terraform-aws-s3/tree/main/examples)
+### Reasons to Use this Module over Stand Alone Resources
+- This module is easy to use with elaborate examples
+- Adheres to AWS security best practices by using checkov to enforce compliance
+- Removes the complexity of configuring multiple stand alone resources
+
```
Comment (hugoalmeida264) [inline README.md:28]: Features missing here, including the massive lifecycle changes but also the security settings you have on the code.

Go through the https://github.com/boldlink/terraform-aws-s3/blob/f5632f80cee906ef15fdb3dd25f76de711be8b9b/CHANGELOG.md for a list of the features of this module ;)

File: README.md [hunk 4]
```diff
@@ -31,13 +36,9 @@ locals {
 }
 
 module "minimum" {
-  source = "../../"
+  source = "boldlink/s3/aws"
```
Comment (hugoalmeida264) [inline README.md]: add `version = "<latest_module_version>"`

File: main.tf [hunk 5]
```diff
@@ -48,11 +48,19 @@ resource "aws_s3_bucket_cors_configuration" "main" {
   }
 }
 
+resource "aws_s3_bucket_ownership_controls" "main" {
```
Comment (hugoalmeida264) [inline main.tf:53]: The ability to set this should be on the `README.md` features list ;)

File: main.tf [hunk 6]
```diff
@@ -48,11 +48,19 @@ resource "aws_s3_bucket_cors_configuration" "main" {
...
+    object_ownership = var.object_ownership
+  }
+}
+
 resource "aws_s3_bucket_acl" "main" {
   count                 = length(keys(var.bucket_acl)) > 0 ? 1 : 0
   bucket                = aws_s3_bucket.main.bucket
   acl                   = try(var.bucket_acl["acl"], null)
   expected_bucket_owner = var.expected_bucket_owner
+  depends_on            = [aws_s3_bucket_ownership_controls.main]
```
Comment (hugoalmeida264) [inline main.tf:65]: Same here

File: main.tf [hunk 7]
```diff
@@ -247,3 +255,141 @@ resource "aws_s3_bucket_replication_configuration" "main" {
   # Must have bucket versioning enabled first
   depends_on = [aws_s3_bucket_versioning.main]
 }
+
+
+resource "aws_s3_bucket_notification" "main" {
+  count       = var.eventbridge || length(var.lambda_function) > 0 || length(var.queue) > 0 || length(var.topic) > 0 ? 1 : 0
```
Comment (hugoalmeida264) [inline main.tf:265]: Well this is also a feature...

File: main.tf [hunk 8]
```diff
@@ -247,3 +255,141 @@ resource "aws_s3_bucket_replication_configuration" "main" {
...
+          date          = try(transition.value.days, null) != null ? null : try(transition.value.date, null)
+          days          = try(transition.value.days, 0)
+          storage_class = transition.value.storage_class
+        }
+      }
+    }
+  }
+
+  # Must have bucket versioning enabled first
+  depends_on = [aws_s3_bucket_versioning.main]
```
Comment (hugoalmeida264) [inline main.tf:398]: And there is no mention on bucket versioning on features

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (patrickmukumbu) [reply CHANGELOG.md:80]: Updated, thanks for that

File: examples/replication_configuration/main.tf (hunk 2 above)
Comment (patrickmukumbu) [reply examples/replication_configuration/main.tf]: Version now locked and exception moved to file

File: README.md (hunk 3 above)
Comment (patrickmukumbu) [reply README.md:28]: Added

File: README.md (hunk 4 above)
Comment (patrickmukumbu) [reply README.md]: Added

File: main.tf (hunk 5 above)
Comment (patrickmukumbu) [reply main.tf:53]: Done

File: main.tf (hunk 6 above)
Comment (patrickmukumbu) [reply main.tf:65]: Added

File: main.tf (hunk 7 above)
Comment (patrickmukumbu) [reply main.tf:265]: Added

File: main.tf (hunk 8 above)
Comment (patrickmukumbu) [reply main.tf:398]: Done

File: README.md [hunk 9]
```diff
@@ -16,10 +16,20 @@
...
-Examples available [here](https://github.com/boldlink/terraform-aws-s3/tree/main/examples)
+### Reasons to Use this Module over Stand Alone Resources
+- This module offers simplicity through detailed examples, making it user-friendly.
+- It follows AWS security best practices by utilizing checkov to ensure compliance.
+- It simplifies the process of setting up s3 bucket with the desired configurations
+- The module provides support for lifecycle configuration.
+- In response to a recent AWS S3 update, this module now includes ownership control, allowing users to specify ownership types. The default is `ObjectWriter`
+- Encryption is enabled by default, allowing users to utilize existing AWS Customer Master Keys (CMKs) or create new ones using the module. Additionally, it supports Server-Side Encryption (SSE).
+- It includes support for S3 bucket notifications, enabling users to set up event-driven workflows.
+- Versioning is enabled by default, providing automatic version control.
```
Comment (hugoalmeida264) [inline README.md]: I would think versioning should be disabled by default, this increases the cost on s3 and not all files are to backup, what do you think?

File: README.md [hunk 10]
```diff
@@ -31,13 +41,10 @@ locals {
...
 module "minimum" {
-  source = "../../"
-  bucket = local.name
-  tags = {
-    Name        = local.name
-    Environment = "Dev"
-  }
-}
+  source  = "boldlink/s3/aws"
+  version = "latest_module_version>"
```
Comment (hugoalmeida264) [inline README.md]: missing `<`

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (patrickmukumbu) [reply README.md]: Added

File: README.md (hunk 9 above)
Comment (patrickmukumbu) [reply README.md]: Yeah, I think this makes sense

Comment (patrickmukumbu) [reply README.md]: I have now disabled it by default, updated README and added checkov exception for the same

File: README.md [hunk 11]
```diff
@@ -16,28 +16,40 @@
...
+- In response to a recent AWS S3 update, this module now includes ownership control, allowing users to specify ownership types. The default is `ObjectWriter`
+- Encryption is enabled by default, allowing users to utilize existing AWS Customer Master Keys (CMKs) or create new ones using the module. Additionally, it supports Server-Side Encryption (SSE).
+- It includes support for S3 bucket notifications, enabling users to set up event-driven workflows.
+- Supports versioning which is in this module disabled by default, providing automatic version control.
+
+Examples available [here](./examples)
 
 ## Usage
-*Points to NOTE*:
+**Points to NOTE**:
```
Comment (hugoalmeida264) [inline README.md]: Change it just to `**NOTE:**`

File: data.tf [hunk 12]
```diff
@@ -0,0 +1,26 @@
+data "aws_caller_identity" "current" {}
+
+data "aws_partition" "current" {}
+
+data "aws_iam_policy_document" "combined" {
+  source_policy_documents = compact([data.aws_iam_policy_document.default.json, var.bucket_policy])
+}
+
+data "aws_iam_policy_document" "default" {
+  statement {
```
Comment (hugoalmeida264) [inline data.tf:10]: @patrickmukumbu why this policy here by default?

File: examples/replication_configuration/main.tf (hunk 2 above)
Comment (hugoalmeida264) [reply examples/replication_configuration/main.tf]: Code still there

File: examples/replication_configuration/main.tf [hunk 13]
```diff
@@ -7,15 +7,22 @@ provider "aws" {
...
 
+resource "random_string" "bucket" {
+  length  = 5
+  special = false
+  upper   = false
+  numeric = false
+}
+
 module "replication_role" {
   #checkov:skip=CKV_TF_1: "Ensure Terraform module sources use a commit hash"
```
Comment (hugoalmeida264) [inline examples/replication_configuration/main.tf]: Remove this exception

**Review by hugoalmeida264 — CHANGES_REQUESTED** [review]

File: data.tf (hunk 12 above)
Comment (patrickmukumbu) [reply data.tf:10]: This was a work around for using `count` condition to create the policy. When part of the policy is dynamically created, in this case the random name....it brought an error that `error.....count attributes cannot be determined until apply` @hugoalmeida264

File: examples/replication_configuration/main.tf (hunk 13 above)
Comment (patrickmukumbu) [reply examples/replication_configuration/main.tf]: Removed

File: README.md (hunk 11 above)
Comment (patrickmukumbu) [reply README.md]: Changed

**Review by hugoalmeida264 — APPROVED** [review]

**Pull request merged**